src/TreeSitter/Importing.hs

```diff
@@ -128,7 +158,56 @@ instance Monad m => Monad (MaybeC m) where
     a' <- a
     case a' of
       Nothing -> pure Nothing
-      Just a -> runMaybeC $ f a
+      Just a  -> runMaybeC $ f a
 
 instance MonadIO m => MonadIO (MaybeC m) where
   liftIO = MaybeC . fmap Just . liftIO
+
+
+newtype FieldName = FieldName { getFieldName :: String }
+  deriving (Eq, Ord, Show)
+
+
+-- | Generic construction of ASTs from a 'Map.Map' of named fields.
+--
+--   Product types (specifically, record types) are constructed by looking up the node for each corresponding field name in the map, moving the cursor to it, and then invoking 'buildNode' to construct the value for that field. Leaf types are constructed as a special case of product types.
+--
+--   Sum types are constructed by attempting to build each constructor nondeterministically. This should instead use the current node’s symbol to select the corresponding constructor deterministically.
+class GBuilding f where
+  gbuildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => Map.Map FieldName Node -> m (f a)
+
+-- Build an empty node of given type
+  gbuildEmpty :: Alternative m => m (f a)
+
+instance (GBuilding f, GBuilding g) => GBuilding (f :*: g) where
+  gbuildNode fields = (:*:) <$> gbuildNode @f fields <*> gbuildNode @g fields
+  gbuildEmpty = (:*:) <$> gbuildEmpty @f <*> gbuildEmpty @g
+
+instance (GBuilding f, GBuilding g) => GBuilding (f :+: g) where
+  gbuildNode fields = L1 <$> gbuildNode @f fields <|> R1 <$> gbuildNode @g fields
+  gbuildEmpty = L1 <$> gbuildEmpty @f <|> R1 <$> gbuildEmpty @g
+
+instance GBuilding f => GBuilding (M1 D c f) where
+  gbuildNode fields = M1 <$> gbuildNode fields
+  gbuildEmpty = M1 <$> gbuildEmpty
+
+instance GBuilding f => GBuilding (M1 C c f) where
+  gbuildNode fields = M1 <$> gbuildNode fields
+  gbuildEmpty = M1 <$> gbuildEmpty
+
+instance (GBuilding f, Selector c) => GBuilding (M1 S c f) where
+  gbuildNode fields =
+    case Map.lookup (FieldName (selName @c undefined)) fields of
+      Just node -> do
+        goto (nodeTSNode node)
+        M1 <$> gbuildNode fields
+      Nothing -> M1 <$> gbuildEmpty
+  gbuildEmpty = M1 <$> gbuildEmpty
+
+instance Building c => GBuilding (K1 i c) where
```

Comment:
This GBuilding instance for constant `K1` calls into the `Building` instance for the field type, which is what lets us treat our special cases above differently (`Maybe`, `[]`, etc.) wherever they occur in the tree.